### PR TITLE
Move responsability of managing Toolbar aspects to the Fragments

### PR DESCRIPTION
**Pull request by hichamboushaba**
### Description
This PR is an attempt to move the responsibility of managing the Toolbar to the fragments, this is useful now, and might be more useful if the team decides to delegate go with Composable Toolbars.
The PR introduces a property `activityAppBarStatus` to the `BaseFragment` class, that returns an `AppBarStatus` instance, and allows child fragments to override it, `MainActivity` uses the `AppBarStatus` instance to decide how to handle the Activity's AppBar and Toolbar, the different options allow handling: visibility, navigation icon, and shadow/divider.

@0nko @JorgeMucientes @hafizrahman since you were involved all in the discussions around the Toolbar, let me know what you think, and whether I should finish the PR or not (if this has any downsides that I'm missing).

### Testing instructions
1. Test the app and confirm the following screens properly hide the Activity Toolbar: "MoreFragment", "ProductImageViewerFragment" and "CouponDetailsFragment".
2. Test the following screens and confirm that the navigation icon is properly set to a `cross` icon: "ProductFilterListFragment" and "IssueRefundFragment"

### Images/gif

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/BaseFragment.kt [hunk 1]
```diff
@@ -16,6 +18,11 @@ open class BaseFragment : Fragment, BaseFragmentView {
         private const val KEY_TITLE = "title"
     }
 
+    open val shouldShowActivityToolbar: Boolean = true
+
+    @DrawableRes
+    open val navigationIconForActivityToolbar: Int? = R.drawable.ic_back_24dp
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/BaseFragment.kt]: This name is way too long, but I didn't find a way to make it shorter while making it clear this impacts only the Activity's Toolbar, I'm afraid if it was just named `navigationIcon`, it might be confusing for someone who added a Toolbar to their Fragment, as they would think it would control its navigation icon.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt [hunk 2]
```diff
@@ -440,26 +447,13 @@ class MainActivity :
...
-                showToolbar()
-            }
-        }
+//        if (isAtRoot) {
+//            toolbar.navigationIcon = null
+//        } else if (showCrossIcon) {
+//            toolbar.navigationIcon = ContextCompat.getDrawable(this, R.drawable.ic_gridicons_cross_24dp)
+//        } else {
+//            toolbar.navigationIcon = ContextCompat.getDrawable(this, R.drawable.ic_back_24dp)
+//        }
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt]: This PR is incomplete, this code would be removed once all the destinations above are migrated to override `navigationIconForActivityToolbar`.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImageViewerFragment.kt [hunk 3]
```diff
@@ -98,6 +102,16 @@ class ProductImageViewerFragment :
         _binding = null
     }
 
+    override fun onStart() {
+        super.onStart()
+        requireActivity().window.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImageViewerFragment.kt]: I moved also the logic of making the screen full screen to the Fragment itself, this would also help clean the MainActivity.

**Comment by JorgeMucientes** [comment]
Thanks for looking into this @hichamboushaba. I like the approach. Moving code out of MainActivity is always great. Moreover IMO this puts the code to handle toolbar visibility in the right place: the fragment (even if the toolbar belongs to the activity). 

Tested it and it works. Haven't found any issue. 

+100 to moving forward with this migration. Nice job.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/BaseFragment.kt (hunk 1 above)
Comment (JorgeMucientes) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/BaseFragment.kt]: The moment you have to add the suffix `ActivityToolbar` to these fields all names are going to be long. But I don't think there is any issue with that. I prefer long names that make it clear what it is than shorter names that make me doubt.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/TopLevelFragment.kt [hunk 4]
```diff
@@ -12,6 +12,9 @@ abstract class TopLevelFragment : BaseFragment, TopLevelFragmentView {
 
     abstract fun shouldExpandToolbar(): Boolean
 
+    override val navigationIconForActivityToolbar: Int?
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/TopLevelFragment.kt]: I don't  think there will ever be a case for what I'm going to say. But having this field as immutable will make it impossible to modify toolbar visibility after Fragment creation. Not that we can modify it now with the current implementation, as is still relying on `onFragmentViewCreated` been invoked. But just thinking out loud if we should try to come up with a more flexible approach that could enable us to easily update toolbar visibility after fragment creation. 
Most likely I'm talking nonsense, so you can ignore this 😅  Just wanted to double-check with the rest if there might be a case for this.

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/TopLevelFragment.kt]: > I don't think there will ever be a case for what I'm going to say

I agree with this 😅, and IMO if there was a case like this, we can then fix it with one of two solutions:
1. Return `false` in `shouldShowActivityToolbar`, and handle the Toolbar completely in the Fragment.
2. Introduce `showActivityToolbar` and `hideActivityToolbar` functions in BaseFragment.

The third option is to make things reactif (maybe using Flow), but IMO this make the API less cleaner, and harder to follow, and I would prefer to avoid this unless there is a real need.

**Comment by hichamboushaba** [comment]
@JorgeMucientes thanks for the initial review, after seeing the new [logic](https://github.com/woocommerce/woocommerce-android/pull/6263) for using a divider instead of shadow in some screens, I decided to update the logic to make the handling of the app bar extensible by using a class instead of plain `Booleans`, so I added a class `AppBarStatus` that allows defining the different options for visibility, navigation icon, shadow, divider, and then MainActivity uses it to update the app bar status.

Please check the code, and if you agree with the approach, then I'll mark the PR as ready for review.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt [hunk 5]
```diff
@@ -149,24 +148,42 @@ class MainActivity :
...
+                AppBarStatus.Hidden -> hideToolbar()
+            }
 
-                // collapsible toolbar should only be able to expand for top-level fragments
-                enableToolbarExpansion(f is TopLevelFragment)
+            if (f is TopLevelFragment) {
+                showBottomNav()
+            } else {
+                hideBottomNav()
             }
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:190]: I moved the logic of the bottom navigation here too, and with this, I got rid of `onDestinationChanged` completely.

**Comment by JorgeMucientes** [comment]
> Please check the code, and if you agree with the approach, then I'll mark the PR as ready for review.
 
Great job on this @hichamboushaba. I like the approach of having an `AppBarStatus` to keep all properties/states of the `toolbar` together. It is also very easy to follow. Also the amount of code removed from `mainActivity` is 😍

I'd say is totally ready for review. I have not approved yet though as I want to test all the screens to double check all behavior is kept as expected. Will look into that today.  Also another pair of eyes would be good as this changes have an impact across the whole app.

**Comment by JorgeMucientes** [comment]
Just finished reviewing all this. Once again, awesome job! 

Just a couple of things I noticed: 

- I think `fun hideToolbar()` from `MainActivity` can be private now
- There is a minor bug althoug it is not related to this changes but I think we can easily fix it in this PR. The bugs is the following, when navigating back and forth between MyStore and MoreMenu the store name subtitle disappears. 

https://user-images.githubusercontent.com/2663464/164705648-0798c995-64c3-4c6f-bafd-3ac8606b0cc1.mp4

This is due to the recent changes done for animating the subtitle visibility between top leve fragment. The problem is `MoreMenu` does not have a visible toolbar resulting in inconsistent visibility states for `toolbarSubtitle`. This patch fixes the issue. 
```diff
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt
index e14b6c4b21..7096dd733b 100644
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt
@@ -36,6 +36,7 @@ import com.woocommerce.android.databinding.ActivityMainBinding
 import com.woocommerce.android.extensions.active
 import com.woocommerce.android.extensions.collapse
 import com.woocommerce.android.extensions.expand
+import com.woocommerce.android.extensions.hide
 import com.woocommerce.android.extensions.navigateSafely
 import com.woocommerce.android.model.Notification
 import com.woocommerce.android.support.HelpActivity
@@ -449,8 +450,12 @@ class MainActivity :
     private fun removeSubtitle() {
         binding.appBarLayout.removeOnOffsetChangedListener(appBarOffsetListener)
         if (binding.toolbarSubtitle.visibility == View.GONE) return
-        binding.toolbarSubtitle.collapse(duration = 200L)
-        hideSubtitleAnimator.start()
+        if (binding.collapsingToolbar.layoutParams.height != 0) {
+            binding.toolbarSubtitle.collapse(duration = 200L)
+            hideSubtitleAnimator.start()
+        } else {
+            binding.toolbarSubtitle.hide()
+        }
     }
```

Feel free to apply that in this PR or let me know if i should create an independent issue for it :)

**Review by JorgeMucientes — APPROVED** [review]

**Comment by hichamboushaba** [comment]
Thanks again @JorgeMucientes.

The issue you mentioned should be fixed in 4aeba4b.
Ready for a last round, and should be ready for merge now.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt [hunk 6]
```diff
@@ -407,15 +408,11 @@ class MainActivity :
...
 
     private fun showToolbar() {
-        if (restoreToolbarHeight > 0) {
-            binding.collapsingToolbar.layoutParams.height = restoreToolbarHeight
-            restoreToolbarHeight = 0
-        }
+        binding.collapsingToolbar.show()
     }
 
     fun hideToolbar() {
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt]: Np this can be private now.

```suggestion
    private fun hideToolbar() {
```

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImageViewerFragment.kt [hunk 7]
```diff
@@ -91,11 +96,21 @@ class ProductImageViewerFragment :
         }
 
         fadeOutToolbarHandler.postDelayed(fadeOutToolbarRunnable, TOOLBAR_FADE_DELAY_MS)
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImageViewerFragment.kt:98]: I've just noticed that with the current implementation we are not showing the fragment's toolbar for a few seconds like we did before. This fragment has its own toolbar, and by a quick look I couldn't tell why this changes break expected behavior. 

This was the behavior prior to the changes: 

https://user-images.githubusercontent.com/2663464/165256895-c3cab39e-5ddb-4305-99d0-49bbc1ae7509.mp4

**Comment by JorgeMucientes** [comment]
Did another round @hichamboushaba 👍🏼 
Left a couple notes. One is super minor, but the other is more concerning.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImageViewerFragment.kt (hunk 7 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImageViewerFragment.kt:98]: Nice catch @JorgeMucientes, I'm glad you caught it.
This should be fixed by the commit [5b80134](https://github.com/woocommerce/woocommerce-android/pull/6255/commits/5b80134b900a814a68956f155a551c0900fa5249), I have no idea what's causing it though, I just found out that hiding the toolbar by setting the height to 0 works, while setting the visibility breaks it.
